### DISCUSSION TITLE
Heterogeneous graph learning with potentially empty node types and edge types

### DISCUSSION THREAD

**Discussion by alvarogunawan** [pyg-team/pytorch_geometric#4143] [Q&A] [answered]
Hi there, I was wondering if there is an expected way of using HeteroData with graphs that contain a node type, but no actual nodes present of that node type. 

The use case here is for multiple heterogeneous graphs which may contain subsets of all the possible node types, e.g. g1 has nodes with node type a, b and c while g2 only has nodes of node type a and b. Likewise for edges, g2 will not have any edges to nodes of type c. These two graphs should both be valid inputs for a single heterogeneous GNN.

**Comment by rusty1s** [comment, marked as answer]
Yes, you might need a workaround to let this use-case work in a mini-batch scenario. It was also reported here: https://github.com/pyg-team/pytorch_geometric/issues/3984.

One simple workaround is to just set the feature matrix to be empty:
```python
data['node_type].x = torch.empty(0, num_features)
```

**Reply by DogitoErgoSum** [reply]
Hi, do i need to do the same to empty edges?
I am doing this but idk if it makes sense

`data["a","to","b"].edge_index=torch.tensor([[],[]])`
where a and b are node types that do not appear on the graph

**Reply by rusty1s** [reply]
Yes, that should work. You should make sure the `dtype` matches though:
```python
torch.tensor([[], []], dtype=torch.long)
```
or
```python
torch.empty((2, 0), dtype=torch.long)
```

**Reply by DogitoErgoSum** [reply]
sorry again, should i do the same for edge_features of nodes tyoes not used in the graph?
